OSDOCS-930 - Include section on the OpenStack Manila CSI Driver (#22199)
openshift/openshift-docs: modules/persistent-storage-csi-manila-dynamic-provisioning.adoc

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent-storage-csi-manila.adoc

[id="persistent-storage-csi-manila-dynamic-provisioning_{context}"]
= Dynamically provisioning Manila CSI volumes

> {product-title} installs a default StorageClass.

Reviewer: This phrase is little bit confusing for me... In OpenShift on OpenStack the default StorageClass `standard` is for Cinder, and it is created by cluster-storage-operator: https://github.com/openshift/cluster-storage-operator/blob/665c4647a815489171eb3f9eff149ae714c98e41/assets/openstack.yaml#L6
Then, there could be several optional StorageClasses for Manila, one for each share type: https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-manila-csi-plugin.md#controller-service-volume-parameters
If I understood it correctly, I suggest to write this line with something like `{product-title} installs a StorageClass for each available Manila share type.`.

Author: This makes more sense, updated.

The YAML files that are created are completely decoupled from Manila and from its Container Storage Interface (CSI) plug-in. As an application developer, you can dynamically provision ReadWriteMany (RWX) storage and deploy Pods with applications that safely consume the storage using YAML manifests.

> 

Reviewer 2: Perhaps add "You can also provision other access modes such as ReadWriteOnce (RWO)."   We cover that at line 38, and we want to emphasize RWX here since lots of other drivers don't have support for that mode, but we don't want to give the impression that somehow this driver can *only* do RWX.

Author: Yeah, I wondered whether it might lead to that impression. Will include your addition.

You can use the same Pod and persistent volume claim (PVC) definitions on-premise that you use with {product-title} on AWS, GCP, Azure, and other platforms, with the exception of the storage class reference in the PVC definition.

.Prerequisites

> * Storage must exist in the underlying infrastructure before it can be mounted as a volume in {product-title}.

Author: Yes, that makes more sense.

> 

Reviewer: There is another prerequisite, that all OCP compute nodes should be connected to the storage network to be able to mount volumes from this network. 
Actually, it's an obvious prerequisite, I don't know if other platforms mention it in their documentation.

Author: We don't mention it elsewhere and our guidance is not to include obvious prereqs, so I'll leave it out.

. In the {product-title} console, click *Storage* → *Persistent Volume Claims*.

. In the persistent volume claims overview, click *Create Persistent Volume Claim*.

. Define the required options on the resulting page.

.. Select the `csi-manila-ceph` StorageClass.

.. Enter a unique name for the storage claim.

.. Select the access mode to specify read and write access for the PVC you are creating.
+
[IMPORTANT]
====
* Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster.

* Use RWO mode if you want to prevent additional Pods from being dynamically provisioned.
====

. Define the size of the storage claim.

. Click *Create* to create the PersistentVolumeClaim and generate a PersistentVolume.

.Procedure (CLI)

To dynamically create a Manila CSI volume using the command-line interface (CLI):

. Create and save a file with the PersistentVolumeClaim object described by the following YAML:

+
.pvc-manila.yaml
[source,yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-manila
spec:
  accessModes: <1>
    - ReadWriteMany
  resources:
    requests:
  	storage: 10Gi

>   storageClassName: slow <2>

Author: Thanks for pointing this out. I received QE feedback to match the source doc I was using, but I too questioned it. I'll change to `gold`.

>   storageClassName: gold <2>

Reviewer: All StorageClasses for Manila are provisioned by the Operator and have the `csi-manila-` prefix: https://github.com/openshift/csi-driver-manila-operator/blob/master/pkg/controller/maniladriver/manila_storageclasses.go#L17

----
+
<1> Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster. Use ReadWriteOnce (RWO) mode to prevent additional Pods from being dynamically provisioned.

> <2> The name of the storage class that provisions the storage back end. This was created by the cluster administrator.

Reviewer: II don't know if I should go into details, but as I said above, all StorageClasses are automatically provisioned by the Manila CSI driver operator. Cluster administrator can create additional classes, but it doesn't make much sense.

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent-storage-csi-manila.adoc

[id="persistent-storage-csi-manila-dynamic-provisioning_{context}"]
= Dynamically provisioning CSI Manila volumes

{product-title} installs a default StorageClass.

As an application developer, you can dynamically provision RWX storage and deploy Pods with applications that safely consume the storage using YAML manifests. These YAML files are completely decoupled from Manila and from its CSI plug-in.

You can use the same Pod and persistent volume claim (PVC) definitions on-premise that you use with {product-title} on AWS, GCP, Azure, and other platforms, with the exception of the storage class reference in the PVC definition.

.Prerequisites

* Storage must exist in the underlying infrastructure before it can be mounted as a volume in {product-title}.

> .UI procedure

Author: I changed it to `Procedure (UI)` and later `Procedure (CLI)` as that's the convention I've seen adopted elsewhere in our docs (e.g., https://docs.openshift.com/container-platform/4.4/storage/persistent_storage/persistent-storage-efs.html#efs-pvc_persistent-storage-efs)

> .UI procedure

Reviewer 3: Aha! Thanks for letting me know what's used throughout :)

To dynamically create a CSI Manila volume using the web console:

. In the {product-title} console, click *Storage* → *Persistent Volume Claims*.

. In the persistent volume claims overview, click *Create Persistent Volume Claim*.

. Define the required options on the resulting page.

.. Select the `csi-manila-ceph` StorageClass.

.. Enter a unique name for the storage claim.

.. Select the access mode to specify read and write access for the PVC you are creating.
+
[IMPORTANT]
====
* Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster.

* Use RWO mode if you want to prevent additional Pods from being dynamically provisioned.
====

. Define the size of the storage claim.

> . Click *Create* to create the PersistentVolumeClaim and generate a PersistentVolume.

Reviewer 4: @jeana-redhat, so are we, honestly.

.CLI procedure

To dynamically create a CSI Manila volume using the command-line interface:

. Create and save a file with the PersistentVolumeClaim object described by the following YAML:

+
.pvc-manila.yaml
[source,yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: myclaim
spec:
  accessModes: <1>
    - ReadWriteMany
  resources:
    requests:
  	storage: 10Gi
  storageClassName: csi-manila-nfs <2>
----
+

> <1> Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster. Use RWO mode to prevent additional Pods from being dynamically provisioned.

Reviewer 3: Can we assume any user reading this knows the ReadWriteMany:RWX abbreviation and others like it?

Author: Not necessarily, but we don't generally write out RWX/ROX/RWO except in the access modes section of Storage docs. IMO it's easy to search for the answer if a newer user isn't familiar. But I'm open to revisiting for standardization, esp. since we try to spell out other storage initialisms, such as PV, PVC, and CSI, on first instance. ¯\_(ツ)_/¯

<2> The name of the storage class that provisions the storage back end. This was created by the cluster administrator.
+

> . Create the object you saved in the previous step by running the following command:

Reviewer 3: ```suggestion
. To create the object that you saved in the previous step, run the following command:
```

Reviewer 4: @jeana-redhat, I've been using this construction, but I'm curious why _To ... run_ is preferable to _Create ... by running_; My original background is not technical writing, so it may be obvious.

Reviewer 3: Disclaimer: I am neither an editor nor an English major, so I may have terms wrong, but will do my best ;)
This suggestion is based on a combination of favoring goal/purpose-oriented statements ("**To accomplish X**, do Y") and using the imperative to describe the action the user should take ("To accomplish X, **do Y**"). "Accomplish X by doing Y" is goal-oriented, but the imperative is on the goal rather than the actual user action. 
Personally, I don't feel there's a huge difference in meaning here, but if you look at the ISG it uses this form consistently (there's an example of this change being explicitly recommended at the end of p 225/start of p 226). There is also of course the additional benefit to translation if we all always use the same forms.

Author: My 2¢: 

Here's another case where minimalism, l10n, and ISG form concentric circles around doing what's best for the user, though with conflicting guidance. 

In this module, the primary goal is stated in the intro: "To dynamically create a CSI Manila volume using the command-line interface...." And I've been habituated to front-load proc sentences with verbs/actions, which is what I see most frequently in our docs ("Click Create to create the instance" instead of "To create the instance, click Create", for example).

Another issue I see with favoring goal-oriented statements is the risk of extraneous info ("Click Save" vs. "To save (?) the file, click Save"). 

Also, I've been forcing myself to use ISG and add "by running the following command(s)", so I'll add that here.

Reviewer 3: Thanks! I of course haven't read enough of our docs to have a feel for what's really common yet :) But consistency is a good thing to add to the balance!

+
----
$ oc create -f pvc-manila.yaml
----
+
A new PVC is created.

> . Verify that the volume was created and is ready by running the following command:

Reviewer 3: ```suggestion
. To verify that the volume was created and is ready, run the following command:
```

Author: I agree with your previous comment that there's not much difference here. I like your construction better here ("run the following command" sounds better than "by running"), so I'll update this line.

+
----
$ oc get pvc pvc-manila
----
+
The `pvc-manila` shows that it is `Bound`.

> You are now ready to use the newly created PVC to configure a Pod.

Reviewer 3: ```suggestion
You can now use the new PVC to configure a Pod.
```
